### PR TITLE
Issue 604 - Entitlement externalKeys over 64 characters allow the creation of duplicates

### PR DESCRIPTION
**Pull request by javier-gomez-github**
- Updated ddl.sql of Account, Tenant, Bundle and Payments.
- Changed max length of the externalKey attributes
- Added exceptions in each API with a new ErrorCode when exceeding the limit
- Added SQL migration files

### CODE REVIEW

File: tenant/src/main/java/org/killbill/billing/tenant/api/user/DefaultTenantUserApi.java [hunk 1]
```diff
@@ -89,8 +90,11 @@ public Tenant createTenant(final TenantData data, final CallContext context) thr
 
         try {
             tenantDao.create(new TenantModelDao(tenant), internalCallContextFactory.createInternalCallContextWithoutAccountRecordId(context));
-        } catch (final TenantApiException e) {
-            throw new TenantApiException(e, ErrorCode.TENANT_CREATION_FAILED);
+        } catch (final Exception e) {
+            if (e.getCause() instanceof SQLDataException) {
+                throw new TenantApiException(e, ErrorCode.EXTERNAL_KEY_LIMIT_EXCEEDED);
+            }
+            else throw new TenantApiException(e, ErrorCode.TENANT_CREATION_FAILED);
```
Comment (sbrossie) [inline tenant/src/main/java/org/killbill/billing/tenant/api/user/DefaultTenantUserApi.java]: Nit: Would be nicer (more java-like) to have:

```
} catch (final TenantApiException e) {
    throw new TenantApiException(e, ErrorCode.TENANT_CREATION_FAILED);
} catch (final  Exception e) {
    if (e.getCause() instanceof SQLDataException) {
        throw new TenantApiException(e, ErrorCode.EXTERNAL_KEY_LIMIT_EXCEEDED);   
    }
    throw e;
} 

```

**Comment by sbrossie** [comment]
Code looks good but i am not seeing exception handling for payment and tenant module (same way was done for account, subscription,..) ?

At least one unit test for one of the module would be nice.

File: subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java [hunk 2]
```diff
@@ -318,7 +319,15 @@ public SubscriptionBaseBundle createBundleForAccount(final UUID accountId, final
...
         final DateTime originalCreatedDate = existingBundles.size() > 0 ? existingBundles.get(0).getCreatedDate() : now;
         final DefaultSubscriptionBaseBundle bundle = new DefaultSubscriptionBaseBundle(bundleKey, accountId, now, originalCreatedDate, now, now);
-        return dao.createSubscriptionBundle(bundle, context);
+        SubscriptionBaseBundle subscriptionBundle = null;
+        try {
+            subscriptionBundle = dao.createSubscriptionBundle(bundle, context);
+        } catch (final Exception e) {
+            if (e.getCause() instanceof SQLDataException) {
+                throw new SubscriptionBaseApiException(e, ErrorCode.EXTERNAL_KEY_LIMIT_EXCEEDED);
+            }
```
Comment (pierre) [inline subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java]: Any other exception is swallowed - you should re-throw instead (same comment elsewhere).

**Comment by pierre** [comment]
> Is there any more precise information you could get from the `SQLDataException` to verify the error comes from the key column?

Actually, since we now enforce a maximum of 255 everywhere, could we simply check the length of the key before going to the DAO?

**Comment by pierre** [comment]
> Actually, since we now enforce a maximum of 255 everywhere, could we simply check the length of the key before going to the DAO?

Going back to the original bug report, you cannot rely on the DAO error anyways since if MySQL is configured in non strict mode, it will happily silently truncate the value.

**Comment by javier-gomez-github** [comment]
@pierre @sbrossie 
Yes, that is what I suggested in my [last comment](https://github.com/killbill/killbill/issues/604#issuecomment-251789773) in the ticket.
I've already finished with the changes and Tests. The PR was updated.
Please feel free to review it.

Thanks!

**Comment by sbrossie** [comment]
@javier-gomez-github Looks good to me. @pierre Any objection if i merge?

File: payment/src/test/java/org/killbill/billing/payment/api/TestPaymentApi.java [hunk 3]
```diff
@@ -868,7 +868,36 @@ public void testCreateSuccessPurchaseWithPaymentControl() throws PaymentApiExcep
...
+        invoice.addInvoiceItem(new MockRecurringInvoiceItem(invoice.getId(), account.getId(),
+                                                            subscriptionId,
+                                                            bundleId,
+                                                            "test plan", "test phase", null,
+                                                            now,
+                                                            now.plusMonths(1),
+                                                            requestedAmount,
+                                                            new BigDecimal("1.0"),
+                                                            Currency.USD));
+
```
Comment (javier-gomez-github) [inline payment/src/test/java/org/killbill/billing/payment/api/TestPaymentApi.java]: You are right, I've just removed it!

File: account/src/test/java/org/killbill/billing/account/api/user/TestDefaultAccountUserApi.java [hunk 4]
```diff
@@ -162,4 +164,16 @@ public void testCreateChildAccountWithInvalidParent() throws Exception {
...
     }
 
+    @Test(groups = "slow", description = "Test Account creation with External Key over limit")
+        public void testCreateAccountWithExternalKeyOverLimit() throws Exception {
+        AccountModelDao accountModelDao = createTestAccount();
+        // Set an externalKey of 256 characters (over limit)
+        accountModelDao.setExternalKey("Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis,.");
+        final AccountData accountData = new DefaultAccount(accountModelDao);
+        try {
+            accountUserApi.createAccount(accountData, callContext);
```
Comment (pierre) [inline account/src/test/java/org/killbill/billing/account/api/user/TestDefaultAccountUserApi.java]: Missing `Assert.fail()`.

File: entitlement/src/test/java/org/killbill/billing/entitlement/api/TestDefaultSubscriptionApi.java [hunk 5]
```diff
@@ -526,6 +525,26 @@ public void testBlockedEntitlementChange() throws AccountApiException, Entitleme
...
+
+        final Account account = createAccount(getAccountData(7));
+
+        final PlanPhaseSpecifier spec = new PlanPhaseSpecifier("Shotgun", BillingPeriod.MONTHLY, PriceListSet.DEFAULT_PRICELIST_NAME, null);
+
+        //2013-08-07
+        final LocalDate effectiveDate = initialDate.plusMonths(1);
+
+        try {
+            entitlementApi.createBaseEntitlement(account.getId(), spec, externalKey, null, effectiveDate, effectiveDate, false, ImmutableList.<PluginProperty>of(), callContext);
```
Comment (pierre) [inline entitlement/src/test/java/org/killbill/billing/entitlement/api/TestDefaultSubscriptionApi.java]: Same here.

File: payment/src/test/java/org/killbill/billing/payment/api/TestPaymentApi.java [hunk 6]
```diff
@@ -868,7 +868,24 @@ public void testCreateSuccessPurchaseWithPaymentControl() throws PaymentApiExcep
...
 
+        final Invoice invoice = testHelper.createTestInvoice(account, now, Currency.USD);
+
+        final String paymentExternalKey = "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis,.";
+        final String transactionExternalKey = "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis,.";
+
+        try {
+            paymentApi.createPurchaseWithPaymentControl(account, account.getPaymentMethodId(), null, requestedAmount, Currency.USD, paymentExternalKey, transactionExternalKey,
+                                                        createPropertiesForInvoice(invoice), INVOICE_PAYMENT, callContext);
+        } catch (final PaymentApiException e) {
```
Comment (pierre) [inline payment/src/test/java/org/killbill/billing/payment/api/TestPaymentApi.java]: Same here.

File: tenant/src/test/java/org/killbill/billing/tenant/api/user/TestDefaultTenantUserApi.java [hunk 7]
```diff
@@ -141,4 +143,18 @@ public void testSystemKeyMultipleValue() throws Exception {
...
+
+    @Test(groups = "slow", description = "Test Tenant creation with External Key over limit")
+    public void testCreateTenantWithExternalKeyOverLimit() throws Exception {
+        final TenantData tenantdata = new DefaultTenant(UUID.randomUUID(),
+                                                        clock.getUTCNow(),
+                                                        clock.getUTCNow(),
+                                                        "Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis,.",
+                                                        "TTR445ee2", "dskjhfs^^54R");
+        try {
+            tenantUserApi.createTenant(tenantdata, callContext);
```
Comment (pierre) [inline tenant/src/test/java/org/killbill/billing/tenant/api/user/TestDefaultTenantUserApi.java]: Same here.

**Comment by javier-gomez-github** [comment]
@pierre 
Done! :shipit:

**Pull request merged**